cmd/nvidia-cdi-hook/cudacompat/cudacompat.go

```diff
@@ -0,0 +1,238 @@
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+**/
+
+package cudacompat
+
+import (
+	"fmt"
+	"os"
+	"path/filepath"
+	"strings"
+
+	"github.com/moby/sys/symlink"
+	"github.com/urfave/cli/v2"
+
+	"github.com/NVIDIA/nvidia-container-toolkit/internal/logger"
+	"github.com/NVIDIA/nvidia-container-toolkit/internal/oci"
+)
+
+const (
+	cudaCompatPath = "/usr/local/cuda/compat"
+)
+
+type command struct {
+	logger logger.Interface
+}
+
+type options struct {
+	hostDriverVersion string
+	containerSpec     string
+}
+
+// NewCommand constructs an cuda-compat command with the specified logger
+func NewCommand(logger logger.Interface) *cli.Command {
+	c := command{
+		logger: logger,
+	}
+	return c.build()
+}
+
+// build the cuda-compat command
+func (m command) build() *cli.Command {
+	cfg := options{}
+
+	// Create the 'cuda-compat' command
+	c := cli.Command{
+		Name:  "cuda-compat",
+		Usage: "This hook ensures that the folder containing the CUDA compat libraries is added to the ldconfig search path if required.",
+		Before: func(c *cli.Context) error {
+			return m.validateFlags(c, &cfg)
+		},
+		Action: func(c *cli.Context) error {
+			return m.run(c, &cfg)
+		},
+	}
+
+	c.Flags = []cli.Flag{
+		&cli.StringFlag{
+			Name:        "host-driver-version",
+			Usage:       "Specify the host driver version. If the CUDA compat libraries detected in the container do not have a higher MAJOR version, the hook is a no-op.",
+			Destination: &cfg.hostDriverVersion,
+		},
+		&cli.StringFlag{
+			Name:        "container-spec",
+			Hidden:      true,
+			Category:    "testing-only",
+			Usage:       "Specify the path to the OCI container spec. If empty or '-' the spec will be read from STDIN",
+			Destination: &cfg.containerSpec,
+		},
+	}
+
+	return &c
+}
+
+func (m command) validateFlags(c *cli.Context, cfg *options) error {
+	return nil
+}
+
+func (m command) run(c *cli.Context, cfg *options) error {
+	s, err := oci.LoadContainerState(cfg.containerSpec)
+	if err != nil {
+		return fmt.Errorf("failed to load container state: %w", err)
+	}
+
+	containerRoot, err := s.GetContainerRoot()
+	if err != nil {
+		return fmt.Errorf("failed to determined container root: %w", err)
+	}
+
+	return m.updateCUDACompatLibs(root(containerRoot), cfg)
+}
+
+func (m command) updateCUDACompatLibs(containerRoot root, cfg *options) error {
+	if !containerRoot.hasPath(cudaCompatPath) {
+		return nil
+	}
+
+	if !containerRoot.hasPath("/etc/ld.so.cache") {
+		// If there is no ldcache in the container, the hook is a no-op.
+		return nil
+	}
+	if !containerRoot.hasPath("/etc/ld.so.conf.d") {
+		// If the /etc/ld.so.conf.d folder does not exist in the container, the hook is a no-op.
+		return nil
+	}
```

Comment:
Question -- what is the rationale behind this condition? I noticed in `createConfig()` we create this directory before attempting to create a file under it.

Reply:
The intent is to not create this in containers that don't already have ldconfig set up to consume config files in that folder. I think we could skip this check though.

Reply:
I have removed this check to simplify the hook.